### PR TITLE
Introduce `AbstractObservable` that represents the location of measurements that should be included in a observable.

### PR DESCRIPTION
**Pull request by inmzhang**
Relates to #271.

I'm no sure whether the `AbstractObservable` representation is convenient for people working on circuits generation. But in principle, this simple representation should include all the information to derive a scalable observable that consists of the measurements.

`AbstractObservable` has two attributes:
1. `top_lines: frozenset[Cube | Pipe]`: the measurements along a line at the top of these blocks should be included in the observable.  
2. `bottom_regions: frozenset[Pipe]`: the stabilizer measurements within a region at the bottom of these block should be included in the observable.

I did not explicitly include the measurement basis or line direction here, because they could be inferred easily from the block type.

Any opinions?

### CODE REVIEW

File: src/tqec/sketchup/block_graph.py
Comment (nelimee) [inline src/tqec/sketchup/block_graph.py:607]: Because having a `None` here should not be possible, I would add an assert instead of a typing cast that has no runtime implication. 
```suggestion
                pipe = self.get_pipe(edge.u.position, edge.v.position)
                assert pipe is not None, f"{edge} is in the graph and should be associated with a Pipe instance."
```

File: src/tqec/sketchup/block_graph.py
Comment (nelimee) [inline src/tqec/sketchup/block_graph.py:612]: Is there a reason why this can only be `v` and not `u`? The underlying type is `nx.Graph` and not `nx.DiGraph`, so at first sight I expect edges to also be undirected (i.e., `u` and `v` are in an arbitrary direction).

File: src/tqec/sketchup/block_graph.py
Comment (nelimee) [inline src/tqec/sketchup/block_graph.py:599]: Same as below:
Because having a None here should not be possible, I would add an assert instead of a typing cast that has no runtime implication.
```suggestion
                cube = self.get_cube(g.nodes[0].position)
                assert cube is not None, f"{g.nodes[0]} is in the graph and should be associated with a Cube instance."
```

File: src/tqec/sketchup/block_graph.py
Comment (nelimee) [inline src/tqec/sketchup/block_graph.py:624]: I do not understand this part. If `surface_attach_direction` is `Direction.Z` that means that we have CUBE1 -> PIPE -> CUBE2 that are all on the same (x, y) coordinates but with increasing `z` right? In which case, why should the pipe be added?

**Review by nelimee — APPROVED** [review]
I left a few questions / comments. I might not be able to re-review before 2 weeks, and even though I did not test the code, I think that it is fine to merge if you tested it yourself and it appears to work. I'll approve to not block that PR during my holidays, but would love to understand the code better before merging.

File: src/tqec/sketchup/block_graph.py
Comment (inmzhang) [reply src/tqec/sketchup/block_graph.py:612]: The pipe's `u`/`v` is always sorted by the position so that `u.position < v.position` in `Pipe.__post_init__` method.

File: src/tqec/sketchup/block_graph.py
Comment (inmzhang) [reply src/tqec/sketchup/block_graph.py:624]: No, here `surface_attach_direction` is `Direction.Z` means that the normal direction of the correlation surface is in the XY plane. Sorry for  the confusing name and I will try to improve that.

**Comment by inmzhang** [comment]
Thank you for the review! I'll keep this PR open for a while to see if anyone who wants to work on the circuit generation part has questions about this representation. Enjoy your holidays!

**Comment by inmzhang** [comment]
I will be taking a two-week vacation to go traveling. During this period, I will keep this PR open.

**Comment by nelimee** [comment]
> Hi Adrien @nelimee, do you need to review this PR again before I merge it?

I will do a quick check again and tell you in less than 30 minutes :)

**Review by nelimee — APPROVED** [review]
LGTM

**Pull request merged**
